libs/core/type_support/include/hpx/type_support/uninitialized_relocate.hpp

```diff
@@ -0,0 +1,127 @@
+//  Copyright (c) 2023 Isidoros Tsaousis-Seiras
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#pragma once
+
+#include <hpx/type_support/is_trivially_relocatable.hpp>
+#include <hpx/type_support/relocate_at.hpp>
+
+#include <cstring>
+#include <type_traits>
+
+namespace hpx {
+
+#if defined(HPX_HAVE_P1144_STD_RELOCATE_AT)
+    using std::uninitialized_relocate;    // what would be the include for this?
+#else
+
+    namespace detail {
+
+        template <typename InIter, typename FwdIter>
+        struct choose_uninitialized_relocate_impl
+        {
+            using object_type =
+                std::decay_t<typename std::iterator_traits<InIter>::reference>;
+            using out_type =
+                std::decay_t<typename std::iterator_traits<FwdIter>::reference>;
```

Comment:
Now that you mention it I am not sure decay is needed at all. 

Would the following be sufficient instead?
```c++
using in_type = typename std::remove_reference_t<decltype(*std::declval<InIter>())>;
using out_type = typename std::remove_reference_t<decltype(*std::declval<FwdIter>())>;
```
 
That way in the case of pointer-to-array types we get `in_type = int[N]`  which we can test against inside `uninitialized_relocate`:

```c+++
template <typename InIter, typename FwdIter>
    FwdIter uninitialized_relocate(InIter first, InIter last, FwdIter dest)
    {
        static_assert(
            hpx::is_relocatable_v<
                typename detail::choose_uninitialized_relocate_impl<InIter,
                    FwdIter>::in_type>,
            "construct_at(dst, std::move(*source)) must be well-formed");
        return detail::uninitialized_relocate_helper(first, last, dest);
    }
```

Reply:
> Would the following be sufficient instead?

Yes, I think so.

But the second half of your comment is coupled to the question above about `relocatable_from`: In _this_ algorithm we _do_ support `first` and `dest` having different `value_type`s (again because it makes the API simpler), so you would need to static-assert on `is_relocatable_from_v<out_type, in_type>`.

Reply:
That's nice, I think is_relocatable_from could be:
```c++
  constexpr bool is_relocatable_from_v =
      std::is_constructible_v<std::remove_cv_t<out_type>,
                              std::add_rvalue_reference_t<in_type>> &&
      !std::is_const_v<in_type>;
```

where `in_type = decltype(*first)`, `out_type = decltype(*dest)`

Reply:
I would say just:
```
constexpr bool is_relocatable_from_v =
    std::is_constructible_v<std::remove_cv_t<out_type>, in_type>;
```
The `add_rvalue_reference_t` isn't needed because all type-traits that talk about invocability ("can I call X with Y arguments?") treat Y in the same way as `declval<Y>()`, i.e., apply a layer of rvalue-ref-ness automatically for no reason. (This is _a priori_ awful, but it's been this way since C++11 and will never change.)

I would expect to have `is_relocatable_from_v<unique_ptr<int>, const unique_ptr<int>> == false`, but `is_relocatable_from_v<shared_ptr<int>, const shared_ptr<int>> == true`.

Reply:
For the first part, does this weird convention apply to constructibility as well? I think this is a counterexample: https://godbolt.org/z/G36nYfeWv

For the second part, I agree with the first example, but I do not understand why `is_relocatable_from_v<shared_ptr<int>, const shared_ptr<int>> == true` would be correct, isn't relocating from a const object UB?

Reply:
> For the first part, does this weird convention apply to constructibility as well? I think this is a counterexample

You must have misinterpreted my meaning, because that's not a counterexample to what I meant; see https://godbolt.org/z/jrx9MMh8T . Basically the intuition is, any time a type-trait _might plausibly_ be implemented as SFINAE on `something-of(declval<Args>()...)`, then it _literally is_ specified as-if-by SFINAE on `something-of(declval<Args>()...)`, i.e., `Args` is interpreted as `Args&&`. Example: https://godbolt.org/z/43rPWTe6T

> For the second part, [...] isn't relocating from a const object UB?

No, I don't see why it should be. Relocation is as-if-by `construct_at(dst, std::move(*src))` (which is OK as long as you've got a copy constructor) followed by `std::destroy_at(src)` (which works fine on const objects). You're not allowed to _change_ the value of a const object, but of course you're allowed to _end its lifetime_. If you couldn't end the lifetime of a const object then every const object would be immortal. :)

Reply:
> You must have misinterpreted ....


Yes, you are completely right, thank you for the clarification

> you're allowed to end its lifetime. If you couldn't end the lifetime of a const object then every const object would be immortal


I had never thought of it that way, thank for this too. TIL